### PR TITLE
make :History show recent files first, and filter duplicates

### PR DESCRIPTION
**Pull request by oconnor663**
Also normalize paths with :~:. These changes make the output of :History
more similar to CtrlP's MRU mode.

### CODE REVIEW

File: autoload/fzf/vim.vim [hunk 1]
```diff
@@ -562,19 +590,25 @@ function! s:format_buffer(b)
...
+" from most-recently-accessed to least-recently-accessed
+function! s:compare_buffers(...)
   let [b1, b2] = map(copy(a:000), 'get(g:fzf#vim#buffers, v:val, v:val)')
   " Using minus between a float and a number in a sort function causes an error
-  return b1 > b2 ? 1 : -1
+  return b1 < b2 ? 1 : -1
+endfunction
+
+" from most-recently-accessed to least-recently-accessed
+function! s:sorted_buffers()
```
Comment (oconnor663) [inline autoload/fzf/vim.vim]: factoring out this function so that `s:history_files()` can use it.

**Comment by junegunn** [comment]
Thanks. We already have `fzf#vim#_uniq` for deduplicating lists. So I believe this should suffice:

```diff
diff --git a/autoload/fzf/vim.vim b/autoload/fzf/vim.vim
index 42df1e0..cbb15b1 100644
--- a/autoload/fzf/vim.vim
+++ b/autoload/fzf/vim.vim
@@ -408,9 +408,11 @@ endfunction
 " History[:/]
 " ------------------------------------------------------------------
 function! s:all_files()
-  return extend(
-  \ filter(reverse(copy(v:oldfiles)), "filereadable(expand(v:val))"),
-  \ filter(map(s:buflisted(), 'bufname(v:val)'), '!empty(v:val)'))
+  return fzf#vim#_uniq(map(
+    \ [expand('%')]
+    \   + filter(map(s:buflisted_sorted(), 'bufname(v:val)'), '!empty(v:val)')
+    \   + filter(copy(v:oldfiles), "filereadable(expand(v:val))"),
+    \ 'fnamemodify(v:val, ":~:.")'))
 endfunction
 
 function! s:history_source(type)
@@ -464,8 +466,8 @@ endfunction
 
 function! fzf#vim#history(...)
   return s:fzf('history-files', {
-  \ 'source':  filter(reverse(s:all_files()), 'v:val != expand("%")'),
-  \ 'options': '-m --prompt "Hist> "'
+  \ 'source':  s:all_files(),
+  \ 'options': ['-m', '--header-lines', '1', '--prompt', 'Hist> ']
   \}, a:000)
 endfunction
 
@@ -565,16 +567,18 @@ endfunction
 function! s:sort_buffers(...)
   let [b1, b2] = map(copy(a:000), 'get(g:fzf#vim#buffers, v:val, v:val)')
   " Using minus between a float and a number in a sort function causes an error
-  return b1 > b2 ? 1 : -1
+  return b1 < b2 ? 1 : -1
 endfunction
 
-function! fzf#vim#buffers(...)
-  let bufs = map(sort(s:buflisted(), 's:sort_buffers'), 's:format_buffer(v:val)')
+function! s:buflisted_sorted()
+  return sort(s:buflisted(), 's:sort_buffers')
+endfunction
 
+function! fzf#vim#buffers(...)
   let [query, args] = (a:0 && type(a:1) == type('')) ?
         \ [a:1, a:000[1:]] : ['', a:000]
   return s:fzf('buffers', {
-  \ 'source':  reverse(bufs),
+  \ 'source':  map(s:buflisted_sorted(), 's:format_buffer(v:val)'),
   \ 'sink*':   s:function('s:bufopen'),
   \ 'options': '+m -x --tiebreak=index --header-lines=1 --ansi -d "\t" -n 2,1..2 --prompt="Buf> "'.s:q(query)
   \}, args)
```

Here I'm using `--header-lines 1` to make the current file unselectable. This happens to work nicely with the current buffer without a name which expands to the name of the working directory.

**Comment by oconnor663** [comment]
Thanks for the quick reply! When I try your version in my home directory, with a new vim instance that hasn't opened anything yet, I find that the header line is actually a file I would've wanted to be selectable. However, if I move to a different directory, then the header line in that case is the cwd like you intended. I think the problem is that `fnamemodify(expand('%'), ":~:.")` from the blank buffer is an empty string if I'm not in a _subdirectory_ of my home folder. So it's blank for `/tmp` and `~`, but not blank for `~/foo`.

**Comment by junegunn** [comment]
Okay, let's just unset `--header-lines` when `expand('%')` is empty. Let me know if you see any issues. Thanks!

**Comment by oconnor663** [comment]
Thank you!

(Merged as e246016844f95a67b9602a2d3e6af24af8dbdbaf, original branch at https://github.com/oconnor663/fzf.vim/tree/mruhistory.)

**Pull request closed without merging**